Unit waypoint overlay

Triangle345 opened the pull request:
The creation of unit waypoint overlay. When you have a unit selected it will now display with a green line where they are going. 
Future ideas 
- designate the type of movement currently instructed to the location (fast move, attack move, regular move etc). 
- Put all factories in a 'factories' empty
- display waypoint overlay on hover

Triangle345 committed:
Terrain debug now displays height

Triangle345 committed:
Merge branch 'master' of https://github.com/FieldWarning/projectFieldWarning

Triangle345 committed:
Added waypoint overlay to moving/selected units

Triangle345 committed:
Cleaned up the commit, organized files and retested

Triangle345 committed:
Merge branch 'master' of https://github.com/FieldWarning/projectFieldWarning

# Conflicts:
#	src/FieldWarning/Assets/Scene/Prefab/Abstract.prefab

Triangle345 committed:
Merge branch 'master' of https://github.com/FieldWarning/projectFieldWarning

Triangle345 committed:
- had to add OverlayFactory back in.. will probably make an 'factories' empty sometime in the future

pvutov commented on src/FieldWarning/Assets/UI/Ingame/SelectionManager.cs:
```diff
@@ -245,18 +254,49 @@ private bool IsInside(Vector3 t)
 
         private void UnselectAll(List<PlatoonBehaviour> selectedPlatoons, bool justPreviewing)
         {
+            foreach (var wp in _selectedWaypointOverlays)
```
I would not use 'var' here, it is not obvious what the type of this variable is.

pvutov commented on src/FieldWarning/Assets/UI/Ingame/SelectionManager.cs:
```diff
@@ -245,18 +254,49 @@ private bool IsInside(Vector3 t)
         {
+            foreach (var wp in _selectedWaypointOverlays)
+            {
+                if (wp != null)
```
Is it actually possible for this to be null? If not, I would leave the check out. I think the overlay is ever destroyed but still in the list, we have some bug that we need to fix.

pvutov commented on src/FieldWarning/Assets/UI/Ingame/SelectionManager.cs:
```diff
@@ -245,18 +254,49 @@ private bool IsInside(Vector3 t)
+                
+
+            }
+
```
That's a lot of random newlines..

pvutov commented on src/FieldWarning/Assets/UI/Ingame/UnitLabel/OverlayFactory.cs:
```diff
@@ -0,0 +1,26 @@
+﻿using PFW.Units;
```
License

pvutov commented on src/FieldWarning/Assets/UI/Ingame/UnitLabel/WaypointOverlayBehavior.cs:
```diff
@@ -0,0 +1,49 @@
+﻿using PFW.Units;
```
License

pvutov commented on src/FieldWarning/Assets/UI/Ingame/UnitLabel/OverlayFactory.cs:
```diff
@@ -0,0 +1,26 @@
+    {
+        
+        var overlayPrefab = Resources.Load<GameObject>("WaypointOverlay");
+        var waypointOverlayBehavior = Object.Instantiate(overlayPrefab, Vector3.zero, Quaternion.identity).GetComponent<WaypointOverlayBehavior>();
```
Line is a bit too long

pvutov commented on src/FieldWarning/Assets/UI/Ingame/UnitLabel/WaypointOverlayBehavior.cs:
```diff
@@ -0,0 +1,49 @@
+    // Update is called once per frame
+    void Update()
+    {
+        if (_platoon != null)
```
I think this null check is a perfect example for the bug hiding thing. We have the bug of waypoints showing after the platoon has been destroyed. If this null check wasn't here, you'd immediately get an error in that case, but instead the code just quietly fails and you're left wondering why.

pvutov commented on src/FieldWarning/Assets/UI/Ingame/SelectionManager.cs:
```diff
@@ -75,7 +79,12 @@ public void Awake()
+        private void Start()
+        {
+            
+        }
```
Doesn't need to be a monobehaviour or have the start method.

pvutov replied:
Should remove the start method especially since this is not a monobehaviour.

pvutov commented on src/FieldWarning/Assets/UI/Ingame/UnitLabel/WaypointOverlayBehavior.cs:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    // Update is called once per frame
+    void Update()
```
private update + delete start

Triangle345 committed:
Cleaned up visual waypoint overlay based on comments from code review.
- Removed mono behavior from SelectionManager and Overlay Factory.
- Added license header to new files
- Added waypoint overlay for queued waypoints

pvutov commented on src/FieldWarning/Assets/UI/Ingame/UnitLabel/OverlayFactory.cs:
```diff
@@ -0,0 +1,49 @@
+{
+    private static OverlayFactory instance; // Needed
+
+    OverlayFactory()
```
public or private?

Triangle345 committed:
Renamed LineCamera to OverlayCamera as this makes a lot more sense.

Triangle345 committed:
- Changed access type of OverlayFactory singleton constructor
- Fixed a race condition by setting a boolean for which instance of a not destroyed duplicate object should be deleted first. Also made the id variable public because it helps us debug
